dilated-file-transfer.md

```diff
@@ -0,0 +1,533 @@
+   --> XXX: see "message encoding" discussion
+
+Control-channel message formats are described using Python pseudo-code to illustrate the data types involved.
+They are actually an encoded `Map` with `String` keys (to use `msgpack` lingo) and values as per the pseudo-code.
+
+All control-channel messages contain a integer "kind" field describing the sort of message it is.
+(That is, `"kind": "text"` for example, not the single-byte tag used for subchannel messages)
+
+The logical "transfer session" lasts as long as the Dilation control-channel is open.
+When either peer is done it closes the control-channel and the mailbox.
+If one side sees the control-channel close this indicates the other peer is done; any outstanding transfers are cancelled.
+Peers SHOULD prefer to cancel individual offers before closing the control channel.
+The mailbox MUST only be closed after the control channel.
+
+Rejected idea: Version message, because we already do version negotiation via mailbox features.
+
+Rejected idea: Offer/Answer messages via the control channel: we need to open a subchannel anyway and the subchannel-IDs are not intended to be part of the Dilation public API.
+
+
+### Control Channel Messages
+
+Each side MAY send a free-form text message at any time.
+These messages look like:
+
+```python
+class Message:
+    message: str     # unicode string
+    kind: str = "text"
+```
+
+
+### Making an Offer
+
+The ability to propose an Offer depends on the `"mode"` of each peer.
+A peer with `"mode": "send"` MAY propose any number of Offers at any time after the connection is established.
+A peer with `"mode": "receive"` MUST NOT propose any Offers.
+A peer with `"mode": "connect"` MAY propose any number of Offers at any time unless the other peer is `"mode": "send"` in which case this peer MUST NOT propose any Offers.
+
+To make an Offer the peer opens a subchannel.
+Recall from the Dilation specification that subchannels are _record_ pipes (not simple byte-streams).
+
+All records on the subchannel begin with a single byte indicating the kind of message.
+Any additional bytes are a kind-dependent payload.
+
+The following kinds of messages exist (as indicated by the first byte):
+* 1: msgpack-encoded `FileOffer` message
+* 2: msgpack-encoded `DirectoryOffer` message
+* 3: msgpack-encoded `OfferAccept` message
+* 4: msgpack-encoded `OfferReject` message
+* 5: raw file data bytes
+* 6: msgpack-encoded `FileAcknowledge` message
+
+All other byte values are reserved for future use and MUST NOT be used.
+
+    XXX: maybe spec [0, 128) as reserved, and [128, 255) for "experiments"?
+
+The peer that opened the new subchannel MUST immediately send one of the two kinds of offer messages.
+
+To offer a single file (with message kind `1`):
+
+```python
+class FileOffer:
+    filename: str   # filename (no path segments)
+    timestamp: int  # Unix timestamp (seconds since the epoch in GMT)
+    bytes: int      # total number of bytes in the file
+```
```

Comment:
Yeah. This is kind of a complex topic and gets a little bit into the "philosophy" here -- that is, this is "a file-transfer protocol" at heart (and not, for example, a "filesystem synchronization protocol" such as `rsync`).

I did consider for example something like encoding `"execute": True` or similar (e.g. "some" of the permissions) but even that seems a bit open for abuse. On the flip side, if you're sending a bunch of shell scripts from one computer to another it might be boring/surprising to have to `chmod +x *.sh` after -- but ultimately safer.

And, as @felinira notes, different platforms have different permissions (and other attributes) available.

These same philosophic considerations come into play with the "symlink" question too -- where the classic protocol takes a more "file transfer" approach versus "synchronize filesystems" (by sending the contents of links).

Using that as guidance, it probably makes sense to _not_ try and transmit permissions? I'm still a little torn about "execute" just from a convenience point of view, but ...overall I think pushing this to "possible future `"feature"`" seems best.